File: challenge02/challenge.js
function getUntilLevel(devs, ids, depth) {
  let _devs = [];

  while (depth > 0) {
    let friends = [];

    for (let i = 0; i < devs.length; i++) {
      const dev = devs[i];

      if (ids.includes(dev.id)) {
        friends = friends.concat(dev.friends);
      }
    }

    depth--;

    if (depth > 0) {
      ids = friends;
    } else {
      _devs = friends;
    }
  }

  return _devs;
}

function unique(arr) {
  return Array.from(new Set(arr));
}

// Using previous values to improve performance
function getSocialBubbles(devs, id) {
  let f1 = unique(getUntilLevel(devs, [id], 1));
  let f2 = unique(getUntilLevel(devs, f1, 1));
  let f3 = unique(getUntilLevel(devs, f2, 1));

  return [f1, f2, f3];
}

// Using generic solution for any depth
function getSocialBubbles02(devs, id) {
  return [
    unique(getUntilLevel(devs, [id], 1)),
    unique(getUntilLevel(devs, [id], 2)),
    unique(getUntilLevel(devs, [id], 3))
  ];
}

const people = [
  { id: "0", friends: ["4", "1"] },
  { id: "1", friends: ["2", "3"] },
  { id: "2", friends: ["1"] },
  { id: "3", friends: ["2", "4"] },
  { id: "4", friends: ["3"] }
];

console.log(">> Social bubble 01 (Jão):", getSocialBubbles(people, "0"));
console.log(">> Social bubble 02 (Jão):", getSocialBubbles02(people, "0"));

File: challenge02/challenge.go
package challenge02

type Person struct {
	ID      string
	Friends []string
}

func unique(s []string) []string {
	inResult := make(map[string]bool)
	var result []string
	for _, str := range s {
		if _, ok := inResult[str]; !ok {
			inResult[str] = true
			result = append(result, str)
		}
	}
	return result
}

func getFriend(devs []Person, id string, depth int) []string {
	var friends []string
	for _, dev := range devs {
		if dev.ID == id {
			friends = dev.Friends
			break
		}
	}

	return friends
}

func GetSocialBubbles(devs []Person, id string) ([]string, []string, []string) {
	var friends []string
	var friendsOfFriends []string
	var friendsOfFriendsOfFriends []string

	for _, dev := range devs {
		if dev.ID == id {
			friends = dev.Friends
			break
		}
	}

	for _, friend := range friends {
		for _, dev := range devs {
			if dev.ID == friend {
				friendsOfFriends = append(friendsOfFriends, dev.Friends...)
				break
			}
		}
	}

	for _, friendOfFriend := range friendsOfFriends {
		for _, dev := range devs {
			if dev.ID == friendOfFriend {
				friendsOfFriendsOfFriends = append(friendsOfFriendsOfFriends, dev.Friends...)
				break
			}
		}
	}

	return friends, unique(friendsOfFriends), unique(friendsOfFriendsOfFriends)
}

// func mapIdsToPersons(ids []string) []*Person {
// 	var output []*Person

// 	for _, dev := range devs {
// 		for _, id := range ids {
// 			if dev.ID == id {
// 				output = append(output, &dev)
// 				break
// 			}
// 		}
// 	}

// 	return output
// }

// func recursiveExtractFriends(devs []*Person, depth int) []string {
// 	friends := []string{}

// 	if (depth == 0) {
// 		return friends
// 	}

// 	for _, dev := range devs {
// 		friends = append(friends, dev.Friends...)
// 	}
// }

// func GetSocialBubbles02(devs []Person, id string) ([]string, []string, []string) {
// 	return recursiveExtractFriends([]*Person{dev}, 3, aux)
// }
